prow/plugins/trigger/trigger.go

```diff
@@ -153,3 +156,91 @@ func isUserTrusted(ghc githubClient, user, trustedOrg, org string) (bool, error)
 	}
 	return orgMember, nil
 }
+
+func runOrSkipRequested(c client, pr *github.PullRequest, requestedJobs []config.Presubmit, forceRunContexts map[string]bool, body, eventGUID string) error {
+	org := pr.Base.Repo.Owner.Login
+	repo := pr.Base.Repo.Name
+	number := pr.Number
+
+	baseSHA, err := c.GitHubClient.GetRef(org, repo, "heads/"+pr.Base.Ref)
+	if err != nil {
+		return err
+	}
+
+	getChanges := fileChangesGetter(c.GitHubClient, org, repo, number)
+	// shouldRun indicates if a job should actually run.
+	shouldRun := func(j config.Presubmit) (bool, error) {
+		if !j.RunsAgainstBranch(pr.Base.Ref) {
+			return false, nil
+		}
+		if j.RunIfChanged == "" || forceRunContexts[j.Context] || j.TriggerMatches(body) {
+			return true, nil
+		}
+		changes, err := getChanges()
+		if err != nil {
+			return false, err
+		}
+		return j.RunsAgainstChanges(changes), nil
+	}
+
+	// For each job determine if any sharded version of the job runs.
```

Comment:
Might be nice to note that sharding here is achieved with presubmit configurations that overlap on name.

Reply:
Done.

Reply:
not up to date on this, are we performing the same logic for postsubmits after this change? we also shard those

Reply:
Postsubmit handling doesn't include any 'Skipped' logic currently so we haven't been seeing problems thus far because we have job shards for all branches. If we had a branch that did not have an associated shard it would never be skipped.
We should update this to 'skip' jobs if needed, but since all the logic in this PR is specific to presubmit jobs I'd prefer to make that change in a follow up PR.

Since we never mark Postsubmit jobs as skipped, at least they are not affected by the bug described in the PR body.
/shrug